Improve License Detection reporting

AyanSinhaMahapatra opened the pull request:
Fixes #3266 
Fixes #3269 

This PR improves the License Detection output with important and cosmetic changes:

- Remove 'not-combined' from the detection log because it is possibly confusing #3266
- Move detection log after matches in the JSON output #3266
- Add rule relevance to the match details to JSON output
- Rename count to detection count in the JSON output
- Show detection_log only when the  --license-text-diagnostics option is requested
- Make license references a command line option, do not include them by default #3269
- Update test expectations accordingly


### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Move detection log after matches in LicenseDetection

Reference: https://github.com/nexB/scancode-toolkit/issues/3266
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove 'not-combined' from LicenseDetection logs

Reference: https://github.com/nexB/scancode-toolkit/issues/3266
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify Misc LicenseDetection attributes

- Rename count to detections_count in UniqueLicenseDetection
- Add rule_relevance back to match level data

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expecatations for LicenseDetection changes

Updates test expectations for a few LicenseDetection cosmatic changes.

- Remove 'not-combined'
- Move detection log after matches
- Add rule relevance
- rename count to detection count

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into license-detection-follow-up

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Only show detection logs on diagnostics option

Only show license detection logs when --license-text-diagnostics is enabled.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expectations for detection logs

Update test expectations after making detection logs only present in
--license-text-diagnostics.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Make license references a CLI option #3269

Moved license references to a CLI option and not default.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regen test expectations after removing license references as default

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on tests/licensedcode/data/additional_licenses/additional_license_combined_test.expected.json:
```diff
@@ -28,6 +26,7 @@
           "matcher": "2-aho",
           "license_expression": "example-installed-2",
           "rule_identifier": "example-installed-2.LICENSE",
+          "rule_relevance": 100,
```
IMHO we should also inline the rule length here, as the combo of 
"score", "matched_length", "match_coverage", "matcher", "rule_relevance" and "rule_length" are what matters to diagnostic detection issues (at least that's what I use)

AyanSinhaMahapatra replied:
The initial idea was, everything that is detection related (example: score, match_coverage) stays here in the match level, everything that is a rule attribute (example: rule_relevance, rule_length) is moved to the license references. Now there are two options:

1.  add a CLI option `inline` which would have all these details, which would facilitate detection diagnostics. 

2. Not have this at all, and we add all the details in the `--review` options, and we have the details available with `--license-reference` anyway.

I think we want to move towards workbench/scancode.io license text to diagnose licence detection issues, and this makes the case where we want to have these details inlined by default a very tiny minority, not sure we want to move away from our general direction to support this. 

What do yoh think @pombredanne?

AyanSinhaMahapatra replied:
After discussion with @pombredanne and @DennisClark, it is concluded that we need to have:

1. lean top level license detections (no match details)
2. resource/package level details have the matches, and also the detection_identifier pointing to the top level unique detection, and not in a separate parallel list.

AyanSinhaMahapatra replied:
This is added, done!

pombredanne commented on tests/licensedcode/data/additional_licenses/additional_license_combined_test.expected.json:
```diff
@@ -386,9 +388,6 @@
           "license_expression": "example-installed-1 AND example-installed-2 AND example1 AND example2 AND apache-2.0",
-          "detection_log": [
-            "not-combined"
-          ],
```
Even if empty, "detection_log" should be there always.

AyanSinhaMahapatra replied:
Refer https://github.com/nexB/scancode-toolkit/issues/3266#issuecomment-1443672136, we have moved detection log to only be there with `--license-text-diagnostics` as we don't want this always.

AyanSinhaMahapatra replied:
Well in the comment it's `--license-diagnostics` but it's weird to have both `--license-diagnostics` and `--license-text-diagnostics` both. We could have the former instead of the later btw, it's probably better (shorter and more general).

AyanSinhaMahapatra replied:
From @pombredanne : we might want to have both, as there could be two different use cases here:
1. having usable matched_text for generating notices, with the detection_diagnostics available.
2. having matched text diagnostics to see what words were matched.

AyanSinhaMahapatra replied:
Done, we have a follow-up at https://github.com/nexB/scancode-toolkit/issues/3288

pombredanne reviewed (changes requested):
LGTM overall. Thanks!
Let's have a chat at your convenience though as I think we can do better wrt. inlining or not inlining license matches.

AyanSinhaMahapatra committed:
Add --license-diagnostics CLI option

Adding a --license-diagnostics options for showing detection logs,
which are now disabled by default. This functionality was added to
--license-text-diagnostics previously, but with this commit it is a
separate CLI option.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add identifier to resource/package license detections

* Top level license detections are lean, without matches
* Resource/package level license detections have the identifier
* identifier is computed scan time, and recomputed in post-processing
  if detections are combined.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regen test expectations after identifier update

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne This is ready for review!

pombredanne commented on tests/licensedcode/data/licenses_reference_reporting/license-reference-works-with-clues.expected.json:
```diff
@@ -3,430 +3,80 @@
-        }
-      ]
+      "detection_count": 1,
+      "detection_log": []
```
I think you may want this removed from top level unless requested.

AyanSinhaMahapatra replied:
@pombredanne false alarm :P This was added because we had `--license-diagnostics` enabled here. 
See for example:

-  a test with this disabled (i.e. not used) [test file](https://github.com/nexB/scancode-toolkit/blob/license-detection-follow-up/tests/licensedcode/test_licenses_reference.py#L32)  [result file](https://github.com/nexB/scancode-toolkit/blob/license-detection-follow-up/tests/licensedcode/data/licenses_reference_reporting/scan-with-reference.expected.json#L83) 

-  a test with this enabled [test_file](https://github.com/nexB/scancode-toolkit/blob/license-detection-follow-up/tests/licensedcode/test_plugin_license_detection.py#L66) [result_file](https://github.com/nexB/scancode-toolkit/blob/license-detection-follow-up/tests/licensedcode/data/plugin_license/unknown_intro/scan-unknown-intro-dual-license.expected.json#L73)

pombredanne reviewed (changes requested):
This is weird... as there should be no detection_log at the top

pombredanne reviewed (commented):
LGTM otherwise and ready to merge!

AyanSinhaMahapatra commented:
Thanks @pombredanne, merging this!
